### PR TITLE
feat: support  remote pinning services in ipfs-http-client

### PR DESCRIPTION
**Pull request by lidel**
> Part of  https://github.com/ipfs/go-ipfs/issues/7559

This adds basic boilerplate for supporting remote pinning commands from https://github.com/ipfs/go-ipfs/pull/7661  in `js-ipfs-http-client`

### TODO

- [x] revisit after https://github.com/ipfs/go-ipfs/pull/7661 switches to https://github.com/ipfs/pinning-services-api-spec/releases/tag/v0.1.2
- [x] ensure commands required by webui integration are mapped: `pin remote add|ls|rm` and `pin remote service add|ls|rm`
- [x] add tests, disable them in js-ipfs for now (only js-ipfs-http-client)

### Refs

Requires: https://github.com/ipfs/go-ipfs/pull/7661
cc: https://github.com/ipfs/ipfs-gui/issues/91, https://github.com/ipfs-shipyard/ipfs-webui/pull/1615, @jessicaschilling, @rafaelramalho19, @achingbrain, @jacobheun

### CODE REVIEW

**Comment by lidel** [comment]
@achingbrain is it possible to add tests to `interface-ipfs-core` but limit implementation to http client for now?

**Comment by achingbrain** [comment]
> is it possible to add generic tests to interface-ipfs-core but limit implementation to http client for now

Yes - add the tests to the suite as normal and skip them in [/packages/ipfs/test/core/interface.spec.js](https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs/test/core/interface.spec.js) and [/packages/ipfs/test/http-api/interface.js](https://github.com/ipfs/js-ipfs/blob/master/packages/ipfs/test/http-api/interface.js)

**Comment by lidel** [comment]
@Gozala mind taking over and pushing this over finish line when you have time?

**Comment by jacobheun** [comment]
@lidel, @hugomrdias might pick up the rest of the pinning service work so he might be able to take this over if @Gozala is full up.

**Comment by lidel** [comment]
Handing over to @hugomrdias :handshake:

File: docs/core-api/PIN.md [hunk 1]
```diff
@@ -25,6 +25,10 @@
   - [Options](#options-4)
   - [Returns](#returns-4)
   - [Example](#example-4)
+- TODO `ipfs.pin.remote.add(ipfsPath, [options])`
+- TODO `ipfs.pin.remote.ls([options])`
+- TODO `ipfs.pin.remote.rm(requestid)`
+- TODO `ipfs.pin.remote.~TBD~replace(requestid, ipfsPath, [options])`
```
Comment (lidel) [inline docs/core-api/PIN.md:31]: Q: does TS make docs automation useful, or do we need to write docs for below ones by hand?

```suggestion
- TODO `ipfs.pin.remote.add`
- TODO `ipfs.pin.remote.ls`
- TODO `ipfs.pin.remote.rm`
- TODO `ipfs.pin.remote.service.add`
- TODO `ipfs.pin.remote.service.ls`
- TODO `ipfs.pin.remote.service.rm`
```

File: packages/interface-ipfs-core/type/pin/remote.ts [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+  name?: string
+
+  /**
+   * Optional list of multiaddrs known to provide the data (max 20).
+   */
+  origins?: Multiaddr[]
+
+  /**
+   * If true, will add to the queue on the remote service and return
+   * RequestID immediately. If false or amitted will wait until pinned on the
```
Comment (lidel) [inline packages/interface-ipfs-core/type/pin/remote.ts:50]: We don't expose it anymore:
```suggestion
   * immediately. If false or amitted will wait until pinned on the
```

File: packages/interface-ipfs-core/type/pin/remote/service.ts [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+   * List registered remote pinning services.
+   */
+  ls(options:ListOptions & AbortOptions):Promise<RemotePinService[]>
+}
+
+export interface Credentials {
+  /**
+   * Service URL
+   */
+  url: URL
```
Comment (lidel) [inline packages/interface-ipfs-core/type/pin/remote/service.ts:26]: Can we rename `url` to `endpoint` to follow [upstream](https://github.com/ipfs/go-ipfs-config/pull/113/files#diff-177fdc882de9eb4633aafaa24834e81a1d30fad53933a66d567a1512ea00014dR18
) fields names?

```json
  "Pinning": {

    "RemoteServices": {
      "test-service": {
        "Api": {
          "Endpoint": "https://testapi.example.com/",
          "Key": "secret"
        }
      },
```


(In the future we may want to add support for libp2p and multiaddrs, and `url` will be awkward)

File: docs/core-api/PIN.md (hunk 1 above)
Comment (achingbrain) [reply docs/core-api/PIN.md:31]: By hand for now - @hugomrdias is working on a types->docs generation tool but it's not ready yet.

**Comment by lidel** [comment]
@Gozala fYSA go-ipfs 0.8.0-rc1 just shipped, you can switch tests to run against https://www.npmjs.com/package/go-ipfs/v/0.8.0-rc1  

https://github.com/ipfs/js-ipfs/blob/4d54a7c207fcfd546d05a743c139937533f455f9/packages/ipfs-http-client/package.json#L83

**Comment by Gozala** [comment]
Errors on CI are related to #3442

**Comment by lidel** [comment]
@Gozala mind rebasing, now that https://github.com/ipfs/js-ipfs/pull/3442 is merged?

**Comment by Gozala** [comment]
> @Gozala mind rebasing, now that #3442 is merged?

I need to rebase it on top of #3449 too, will have it ready today

**Comment by Gozala** [comment]
I would like to add few more tests, but I think it's ready for the review.

**Comment by lidel** [comment]
@Gozala  go-ipfs 0.8.0-rc2 will include https://github.com/ipfs/go-ipfs/pull/7829 – will it "just work" or do we need to update this PR?

**Comment by lidel** [comment]
@Gozala also, had trouble with building unless I bumped "ipfs-core-types" to  "^0.2.0" – mind rebasing this PR?

**Comment by Gozala** [comment]
For some reason when testing in browser node process just won't exit. It must be due to https://github.com/ipfs/js-ipfs/pull/3293/files#diff-bd8c36bcaf63bb93ceec2bb8a8a7d261692b8e416c10eed90d3e03232bc3873e but I just can't seem to figure out why.

**Review by achingbrain — APPROVED** [review]
LGTM, just need to sort the merge conflicts out

**Comment by lidel** [comment]
@Gozala FYI I've just bumped this PR to use go-ipfs 0.8.0-rc2 so we identify issues early (if any)

**Pull request merged**